man/man3/angle_lib.3
.\" 
.\" Copyright (c) 2011, Gordon D. Carrie. All rights reserved.
.\" 
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 
.\"     * Redistributions of source code must retain the above copyright
.\"     notice, this list of conditions and the following disclaimer.
.\"     * Redistributions in binary form must reproduce the above copyright
.\"     notice, this list of conditions and the following disclaimer in the
.\"     documentation and/or other materials provided with the distribution.
.\" 
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
.\" "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
.\" LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
.\" A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
.\" HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
.\" SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
.\" TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
.\" PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
.\" LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
.\" NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
.\" SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\" 
.\" Please address questions and feedback to dev0@trekix.net
.\" 
.\" $Revision: 1.5 $ $Date: 2011/11/28 16:43:52 $
.\"
.TH angle_lib 3 "angle functions"
.SH NAME
Angle_AzWrap, Angle_ElevWrap, Angle_GCDist \- angle functions
.SH SYNOPSIS
.nf
\fB#include "angle_lib.h"\fP
\fBdouble Angle_AzWrap(const double\fP \fIaz\fP, \fBconst double\fP \fIr\fP\fB);\fP
\fBdouble Angle_ElevWrap(const double\fP \fIelev\fP\fB);\fP
\fBdouble Angle_GCDist(const double\fP \fIaz1\fP, \fBconst double\fP \fIelev1\fP,
	\fBconst double\fP \fIaz2\fP, \fBconst double\fP \fIelev2\fP\fB);\fP
\fBdouble Angle_GCAz(const double\fP \fIaz1\fP, \fBconst double\fP \fIelev1\fP,
	\fBconst double\fP \fIaz2\fP, \fBconst double\fP \fIelev2\fP\fB);\fP
\fBvoid Angle_GCStep(const double\fP \fIaz\fP, \fBconst double\fP \fIelev\fP,
	\fBconst double\fP \fIdirn\fP, \fBconst double\fP \fIdist\fP, \fBdouble\fP *az1\fP, \fBdouble\fP *\fIelev1\fP\fB);\fP
.fi
.SH DESCRIPTION
In this context, angles az1 and az2 refer to azimuth, a.k.a. longitude or phi, and
elev1 and elev2 refer to elevation, also known as latitude or theta.
Angles are expressed in radians.

\fBAngle_AzWrap\fP returns \fIaz\fP\ +\ i\ *\ \fBpi\fP for some integer i, in the
interval [\fIr\fP\ -\ \fBpi\fP\ /\ 2,\ \fIr\fP\ +\ \fBpi\fP\ /\ 2).

\fBAngle_ElevWrap\fP returns the elevation (latitude) reached by going \fIelev\fP
radians up (North) from the plane of \fB0\fP elevation (the Equator).

\fBAngle_GCDist\fP returns the great circle distance between points
(\fIaz1\fP,\ \fIelev1\fP) and (\fIaz2\fP,\ \fIelev2\fP).

\fBAngle_GCAz\fP returns the angle from North clockwise to the great circle
connecting point (\fIaz1\fP,\ \fIelev1\fP) to point (\fIaz2\fP,\ \fIelev2\fP).

\fBAngle_GCStep\fP computes the azimuth and elevation of the point in direction
\fIdirn\fP and distance \fIdist\fP along a great circle from point
(\fIaz\fP,\ \fIelev\fP).  It places the azimuth and elevation of the destination
point at addresses \fIaz1\fP and \fIelev1\fP respectively.
.SH BUGS
\fBAngle_AzWrap\fP is sensitive to round off errors when
\fB|\fP\fIl\fP\ \fB-\fP\ \fIr\fP\fB|\fP\ \fB=\ pi\fP.  For example,
if \fIl\fR\ =\ \fB-170.0\fP degrees, \fIr\fP\ =\ \fB10.0\fP degrees, then with
conversions to and from radians, \fBAngle_AzWrap(l, r)\fP returns \fB-170.0\fP
degrees on some platforms and \fB190.0\fP degrees on others.
.SH AUTHOR
Gordon Carrie (dev0@trekix.net)


src/angle_lib.c
/*
   -	angle_lib.c --
   -		This file defines functions that manipulate angles.
   -		See angle_lib (3).
   -
   .	Copyright (c) 2011, Gordon D. Carrie. All rights reserved.
   .	
   .	Redistribution and use in source and binary forms, with or without
   .	modification, are permitted provided that the following conditions
   .	are met:
   .	
   .	    * Redistributions of source code must retain the above copyright
   .	    notice, this list of conditions and the following disclaimer.
   .
   .	    * Redistributions in binary form must reproduce the above copyright
   .	    notice, this list of conditions and the following disclaimer in the
   .	    documentation and/or other materials provided with the distribution.
   .	
   .	THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
   .	"AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
   .	LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
   .	A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
   .	HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
   .	SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED
   .	TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
   .	PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
   .	LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
   .	NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
   .	SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
   .
   .	Please send feedback to dev0@trekix.net
   .
   .	$Revision: 1.4 $ $Date: 2011/11/28 16:43:52 $
 */

#include <math.h>
#include "angle_lib.h"

double Angle_AzWrap(const double l, const double r)
{
    double l1 = fmod(l, 2.0 * M_PI);
    l1 = (l1 < r - M_PI) ? l1 + 2.0 * M_PI
	: (l1 >= r + M_PI) ? l1 - 2.0 * M_PI : l1;
    return (l1 == -0.0) ? 0.0 : l1;
}

double Angle_ElevWrap(const double l)
{
    double l1 = fmod(l, 2.0 * M_PI);
    l1 += (l1 < 0.0) ? 2.0 * M_PI : 0.0;
    return (l1 > 1.5 * M_PI) ? l1 - 2.0 * M_PI : (l1 > M_PI_2 ) ? M_PI - l1 : l1;
}

double Angle_GCDist(const double o1, const double a1,
	const double o2, const double a2)
{
    double sin_do_2, sin_da_2, a;

    /*
       Reference -- R.W. Sinnott, "Virtues of the Haversine",
       Sky and Telescope, vol. 68, no. 2, 1984, p. 159
       cited in: http://www.census.gov/cgi-bin/geo/gisfaq?Q5.1
     */

    sin_do_2 = sin(0.5 * (o2 - o1));
    sin_da_2 = sin(0.5 * (a2 - a1));
    a = sqrt(sin_da_2 * sin_da_2 + cos(a1) * cos(a2) * sin_do_2 * sin_do_2);
    return (a > 1.0 ? M_PI : 2.0 * asin(a));
}

double Angle_GCAz(const double o1, const double a1,
	const double o2, const double a2)
{
    double sin_da, sin_sa, y, x;

    sin_da = sin(a1 - a2);
    sin_sa = sin(a2 + a1);
    y = cos(a2) * sin(o2 - o1);
    x = 0.5 * (sin_sa - sin_da - (sin_sa + sin_da) * cos(o2 - o1));
    return atan2(y, x);
}

void Angle_GCStep(const double o1, const double a1, const double d, const double s,
	double *o2, double *a2)
{
    double sin_s, sin_d, cos_d, dlon, a, x, y;

    /*
       Reference -- Smart, W. M., "Textbook on Spherical Astronomy",
       Sixth edition revised by R. M. Green.
       Cambridge University Press, Cambridge, 1977.
     */

    sin_s = sin(s);
    sin_d = sin(d);
    cos_d = cos(d);
    a = 0.5 * (sin(a1 + s) * (1.0 + cos_d) + sin(a1 - s) * (1.0 - cos_d));
    *a2 = (a > 1.0) ? M_PI_2 : (a < -1.0) ? -M_PI_2 : asin(a);
    y = sin_s * sin_d;
    x = 0.5 * (cos(a1 + s) * (1 + cos_d) + cos(a1 - s) * (1 - cos_d));
    dlon = atan2(y, x);
    *o2 = Angle_AzWrap(o1 + dlon, 0.0);
}
